h2o-algos/src/test/java/hex/grid/GridSearchWithFilteringTest.java

```diff
@@ -0,0 +1,137 @@
+        filteredGridItems.add(gbmParams);
+      }
+    }
+
+    assertEquals(totalNumberOfPermutationsInHPGrid - expectedNumberOfFilteredOutPermutations, filteredGridItems.size());
+  }
+
+  @Test
+  public void cartesianGridSearchWithFunctionsToFilterOutUnnecessaryGridItems_GBMParametersGrid() {
+    Map<String, Object[]> hpGrid = new HashMap<>();
+    hpGrid.put("_max_depth", new Integer[]{3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17});
+    hpGrid.put("_min_rows", new Integer[]{1, 5, 10, 15, 30, 100});
+    
+    int totalNumberOfPermutationsInHPGrid = 90;
+
+    GBMModel.GBMParameters gbmParameters = new GBMModel.GBMParameters();
+
+    GridSearch.SimpleParametersBuilderFactory simpleParametersBuilderFactory = new GridSearch.SimpleParametersBuilderFactory();
+
+    HyperSpaceSearchCriteria.CartesianSearchCriteria hyperSpaceSearchCriteria = new HyperSpaceSearchCriteria.CartesianSearchCriteria();
+
+    PermutationFilterFunction strictFilterFunction = gridItem -> {
+      return !((int) gridItem.get("_max_depth") >= 15.0 && (int) gridItem.get("_min_rows") >= 30);
+    };
+    int expectedNumberOfFilteredOutPermutations = 6;
+
+    ArrayList<PermutationFilterFunction> filterFunctions = new ArrayList<>();
+    filterFunctions.add(strictFilterFunction);
+
+    HyperSpaceWalker.CartesianWalker<GBMModel.GBMParameters> walker =
+            new HyperSpaceWalker.CartesianWalker<>(gbmParameters,
+                    hpGrid,
+                    simpleParametersBuilderFactory,
+                    hyperSpaceSearchCriteria,
+                    filterFunctions);
+
+    HyperSpaceWalker.HyperSpaceIterator<GBMModel.GBMParameters> iterator = walker.iterator();
+
+    ArrayList<GBMModel.GBMParameters> filteredGridItems = new ArrayList<>();
+    while (iterator.hasNext(null)) {
+      GBMModel.GBMParameters gbmParams = iterator.nextModelParameters(null);
+      if( gbmParams != null) { // we might have had next element ( iterator.hasNext) = true) but it could be filtered out by filtering functions
+        filteredGridItems.add(gbmParams);
+      }
+    }
+
+    assertEquals(totalNumberOfPermutationsInHPGrid - expectedNumberOfFilteredOutPermutations, filteredGridItems.size());
+  }
+
+  @Test
+  public void test_class_KeepOnlyFirstMatchFilterFunction_was_extended_properly() {
+    KeepOnlyFirstMatchFilterFunction ff1 = new KeepOnlyFirstMatchFilterFunction((gridItem) -> {
+      Object mainHP = gridItem.get("_blending");
+      return mainHP instanceof Boolean && !(Boolean) mainHP;
+    });
+
+    KeepOnlyFirstMatchFilterFunction ff2 = new KeepOnlyFirstMatchFilterFunction((gridItem) -> {
+      Object mainHP = gridItem.get("_noise_level");
+      return mainHP instanceof Double && (Double) mainHP == 0.01;
+    });
+
+    Map<String, Object> permutation = new HashMap<>();
+    permutation.put("_blending", false);
+    permutation.put("_noise_level",  0.01);
+    
+    assertTrue(ff1.apply(permutation));
```

Comment:
how does this test any of the functionality when each function is called only once in the test?

Reply:
@honzasterba test can test any assumption. It is I believe totally fine that someone tests correct usage of library's API and functionality especially if he employs TDD approach. But I agree that later( when code is written) sometimes it does not add much value. Will remove this one.

Reply:
my issue is that this test would pass even if you remove the pass-only-once wrapper, and as such its not testing the the thing it is supposed to be testing, by all means keep it but add another assert